src/bson.c

```diff
@@ -1007,6 +1007,7 @@ PHP_FUNCTION(toArray)
 
 	MAKE_STD_ZVAL(state.zchild);
 	if (!bson_to_zval((const unsigned char *)data, data_len, &state)) {
+		zval_ptr_dtor(&state.zchild);
```

Comment:
`MAKE_STD_ZVAL(state.zchild)` is always called before we venture into `bson_to_zval()`, which is why we free. This shouldn't depend on whether we called `array_init()` within `bson_to_zval()`.